[Rules] [Feature Request] Adding support for substring and hex parsing

crathje opened the pull request:
Adding support for
* substring
* strtol (useful for converting hex or binary numbers)
* 100ths division
in eventvalue parsing.

Using this modification I am able to handle the messages sent from Opentherm Gateway ( http://otgw.tclcode.com/ ) and to set variables of a dummy device in rules.

Example:
* Message coming from the serial interface: **T101813C0**
  * The B denotes that the message is from the 
  * The next 4 bytes (actually 2bytes hex encoded) denote the status and type of the message.
  * the last 4 bytes (actually 2bytes hex encoded) denote the payload.
* Message that ends up in rules when using ser2net (P020) and Generic handling: **!Serial#BT101813C0**

The room temperature in this sample is 19.75°C

To get this I need to access the last four bytes in packs of two bytes:
`[substring:13:15:%eventvalue%]`
and
`[substring:15:17:%eventvalue%]`

Parsing them to decimal representation each (using a base 16 call to strtol):
`[strtol:16:[substring:13:15:%eventvalue%]]`
and 
`[strtol:16:[substring:15:17:%eventvalue%]]`

Last but not least the fraction is not correct, it needs to be divided by 256 (and multiplied by 100) with a call to div100ths:
`[div100ths:[strtol:16:[substring:15:17:%eventvalue%]]:256]`

And now a complete rule that I use to parse this and set a variable in a dummy device:
```
// Room temperature
on !Serial#T1018* do
TaskValueSet 2,1,[strtol:16:[substring:13:15:%eventvalue%]].[div100ths:[strtol:16:[substring:15:17:%eventvalue%]]:256]
endon
```
Now the variable can be used in the rules/controller/log/whatever.

crathje committed:
Adding support for
* substring
* strtol (useful for converting hex or binary numbers)
* 100ths division
in eventvalue parsing.

Using this modification I am able to handle the messages sent from Opentherm Gateway ( http://otgw.tclcode.com/ ) and to set variables of a dummy device in rules.

Example:
* Message coming from the serial interface: T101813C0
** The B denotes that the message is from the boiler.
** The next 4 bytes (actually 2bytes hex encoded) denote the status and type of the message.
** the last 4 bytes (actually 2bytes hex encoded) denote the payload.
* Message that ends up in rules when using ser2net (P020) and Generic handling: !Serial#BT101813C0

The room temperature in this sample is 19.75°C

Rule used to parse this and set a variable in a dummy device:
// Room temperature
on !Serial#T1018* do
TaskValueSet 2,1,[strtol:16:[substring:13:15:%eventvalue%]].[div100ths:[strtol:16:[substring:15:17:%eventvalue%]]:256]
endon

Now the variable can be used in the rules/controller/log/whatever.

TD-er commented on src/ESPEasyRules.ino:
```diff
@@ -304,28 +304,84 @@ void replace_EventValueN_Argv(String& line, const String& argString, unsigned in
+    idxEnd = substCmd.indexOf("]", idxStart);
+    String stringValue = substCmd.substring(idxStart, idxEnd);
+  
+    line.replace(substCmd, stringValue.substring(substStart.toInt(), substEnd.toInt()));
```
You don't check here if the `substStart` or `substEnd` is actually an int.
See: https://github.com/letscontrolit/ESPEasy/blob/0c6e6ca915e5440de54213246dc8c1a8e6ca9cce/src/Misc.ino#L1320-L1327

crathje replied:
I just added the checks in a more recent commit: 
https://github.com/crathje/ESPEasy/commit/d238593ac0b039cdc10a0b920de9a3fb20f128e7

TD-er commented on src/ESPEasyRules.ino:
```diff
@@ -304,28 +304,84 @@ void replace_EventValueN_Argv(String& line, const String& argString, unsigned in
+       
+    line.replace(strolCmd, String(strtol(stringValue.c_str(), NULL, base.toInt())));   
+  }
+}
```
This function is almost the same as `process_substring` and `process_div100ths`.
Maybe you can make a function to determine the ranges?
Just to make sure we don't have a lot of code duplication.

crathje replied:
I joined the functions in https://github.com/crathje/ESPEasy/commit/d238593ac0b039cdc10a0b920de9a3fb20f128e7

TD-er commented on src/ESPEasyRules.ino:
```diff
@@ -304,28 +304,84 @@ void replace_EventValueN_Argv(String& line, const String& argString, unsigned in
+  // process other markups as well
+  process_substring(line);
+  process_strtol(line);
+  process_div100ths(line);
```
Just curious. Why do these markup parsing only in `substitute_eventvalue` and not in `parseTemplate` ?
https://github.com/letscontrolit/ESPEasy/blob/0c6e6ca915e5440de54213246dc8c1a8e6ca9cce/src/Misc.ino#L1538

The parseTemplate is also used on all other uses where we may address values from other tasks or system variables.

crathje replied:
To be honest? I downloaded the source of ESPEasy like the day I issued the pull request and still did not figure out the whole control flow.

crathje committed:
Fixed wildcard matching which only checked up to one character before the actual wildcard and not up to the wildcard.

crathje committed:
Added checks for valid integers as suggested by TD-er
Merged the functions as suggested by TD-er

crathje reviewed (commented):
Just in case it gets dismissed: I fixed another issue regarding the wildcard matching in https://github.com/crathje/ESPEasy/commit/031888f02411087c550ae25d3692f5020bd4c1a2

crathje committed:
Added [ord: to get integer value of a string/character.

TD-er commented on src/ESPEasyRules.ino:
```diff
@@ -304,59 +304,58 @@ void replace_EventValueN_Argv(String& line, const String& argString, unsigned in
 
-void process_strtol(String& line) {
-  // Syntax like 1234[strtol:16:38]7890
+void process_internal(String &line, String cmd_s) {
```
Please use `const String& cmd_s` as parameter.
Now you're making a deep copy.

TD-er commented on src/ESPEasyRules.ino:
```diff
@@ -304,59 +304,58 @@ void replace_EventValueN_Argv(String& line, const String& argString, unsigned in
-    float val = (100.0 * dividend.toInt()) / (1.0 * divisor.toInt());
+  int iarg1, iarg2;
+  while ((startIndex = line.indexOf(cmd_s)) != -1) {
+    String fullCommand = line.substring(startIndex, line.indexOf("]", startIndex) + 1);
```
Maybe also check for the presence of `]` in the string?
For example:
```c++
  while ((startIndex = line.indexOf(cmd_s)) != -1 && line.indexOf("]", startIndex) != -1) {
```

crathje replied:
Added the check.

TD-er commented on src/ESPEasyRules.ino:
```diff
@@ -304,59 +304,58 @@ void replace_EventValueN_Argv(String& line, const String& argString, unsigned in
+          && validIntFromString(arg2, iarg2)) {
+        replacement = arg3.substring(iarg1, iarg2);
+      }
+    }
```
These `if (cmd_s.equalsIgnoreCase...` can be `if ... else` statements instead of `if ... if ... etc.`
And also you can copy the cmd_s to a new String and convert it to lower case.
Then you don't need to check case insensitive, which is quite an expensive operation.

crathje replied:
switched to if ... else if...

crathje replied:
Copied string to lowercase to avoid checking for case insenstive variant.

crathje committed:
Avoiding deepcopy on command pass.
Checking for closing character ].
Switched to if else if.
Copied command string to get a lowercase representation.

TD-er committed:
[docs] Document string handling in rules (PR #2828 )

Closes: #2828

TD-er committed:
Merge branch 'mega' into mega

TD-er committed:
Add note to optimize code

TD-er commented:
I fixed a merge conflict, created documentation and will very likely merge it tomorrow.

For any future pull request you may want to make.
Please create a branch on your own fork on which you will commit your changes and then create a pull request for that branch.
Now you were working on the main branch and if you would later had made any changes, those would also have been included in this pull request.

TD-er commented:
OK, I'm now testing it a bit more and these transforms do not work well on `[taskname#varname]`, so I will try to fix those as well.

TD-er commented:
Hmm this looks like it is causing a major rewrite of the rules parsing engine.
The reason is we now parse a line from left to right and replace whatever we come across, while your new commands require a more depth-first approach.
The parameters you need are in the back of the command, or even nested.

So what should be done here is something like this:  (just written here as memory dump, not tested)

```c++
String parseline(const String& input, int start, int& end, bool useURLencode) {
  int newStart = input.indexOf('[', start + 1);
  if (newStart != -1) {
    int newEnd = end;
    String subResult = parseline(input, newStart, newEnd, useURLencode);
    int sectionEnd = input.indexOf(']', newEnd);
    if (sectionEnd != -1) {
      // Found a closing bracket, so we have to interpret the content of this section.
      end = sectionEnd;
      String section = input.substring(start, newStart);
      section += subResult;
      section += input.substring(newEnd, sectionEnd);
      // interpret content
      return parseTemplate_padded(section, useURLencode);
    }
    addLog(LOG_LEVEL_ERROR, F("Error: bracket mismatch");
  }
  return input.substring(start, end);
}
```

This is just a dump of what I'm now thinking of and it does also need quite a bit of rework on the parseTemplate function as well, since we're essentially making the parser quite a bit more flexible.
But the problem is that recursive calls do need more stack and that's maybe already quite limited and may lead to crashes.
Also it does need quite a bit of String allocations.

N.B. this suggested `parseline` should also check if there is another opening bracket left, so it needs a while loop and also it does not take into account it can already be the inner part, so this will not work yet, but I think you will get the idea.


Edit:
Made an issue for it: https://github.com/letscontrolit/ESPEasy/issues/2899

TD-er committed:
[Rules] Make nested string modification commands possible

TD-er commented:
Just couldn't let it go, so I just made it possible to use nested string commands.

So this is now also possible:
```
on DS-1#Temperature do
  logentry,[substring:0:2:[strtol:16:[substring:0:2:[DS-1#Temperature]][substring:3:5:[DS-1#Temperature]]]]
endon
```

And some debug log for illustrative purposes:
```
 221313 : Info  : EVENT: DS-1#Temperature=22.13
 221346 : Info  : parse_string_commands cmd: substring:0:2:22.13 -> 22
 221347 : Info  : parse_string_commands cmd: substring:3:5:22.13 -> 13
 221348 : Info  : parse_string_commands cmd: strtol:16:2213 -> 8723
 221349 : Info  : parse_string_commands cmd: substring:0:2:8723 -> 87
 221350 : Info  : ACT  : logentry,87
 221351 : Info  : Command: logentry
 221353 : Info  : 87
```

giig1967g commented:
@TD-er 
Are you sure this is the right approach?
We are adding another layer of syntax in the rules... a layer dedicated to a single device (will all respect to the OP, who will ever use the "div100ths" command?)
And maybe for just a change that helps a single device, we could have impact on memory, speed and adding potential bugs and complexity.
Also, if you go for this approach, you will need to potentially create commands inside the core for each new device on request.

Probably a more general solution should be to create a "Opentherm Gateway" plugin.
And maybe a REDEX format or command for manipulating strings.

These are my 2 cents, of course...

TD-er commented:
I was also a bit confused about this one: `div100ths`, and also thought about maybe making it one with 2 parameters (100 and 255 of this example).
But it is something that you may see in several sensors, but not always limited to a single byte.
Also the 2 decimals output is a bit strange here.

So in that respect I do share your concern about this one.

The pro for such a command is to make sure the result is a float in between steps, or else you will loose a lot of resolution.

For now, I will remove this specific one (comment out) as it may indeed lead to a lot of these.

The reason I put some work in this was also because I suddenly thought of a way to make parsing with nested brackets without using recursive calls and lots of memory allocation.
That may also be used to parse other variable notations (on which you also contributed in the past) and speed up rules where some expressions are used several times.

Apart from that there have been a number of requests to split longer numbers, like in the processing of RFID etc, so that was also very nice to see implemented by this PR.

But thanks for the critical look on the div100ths, as that's a bit too specific indeed.

uzi18 commented:
@TD-er  maybe change [ ] to {} or <>

TD-er commented:
> @TD-er maybe change [ ] to {} or <>

Just curious, why?

uzi18 commented:
because we use these chars already for something else and now your example line is completely unreadable

TD-er commented:
That's a valid point.
On the other hand we also have some formatting options on `[task#var]`
Using {} then? <> is already used in compares.

uzi18 commented:
{} left untouched
we should extend these formating options to support substring etc. to have everything consistent

maybe later we can add some user variables dedicated for strings manipulation?

TD-er commented:
Well the % notation I made for some standard conversions are also a bit unreadable, so maybe something like `[c_to_f;<value>]` may be more clear.
And in your suggestion, using {}
